examples/large_models/vllm/base_vllm_handler.py

```diff
@@ -21,59 +24,64 @@ def __init__(self):
         self.initialized = False
 
     def initialize(self, ctx):
-        ctx.cache = {}
-
         self.model_dir = ctx.system_properties.get("model_dir")
         vllm_engine_config = self._get_vllm_engine_config(
             ctx.model_yaml_config.get("handler", {})
         )
         self.adapters = ctx.model_yaml_config.get("handler", {}).get("adapters", {})
-        self.vllm_engine = LLMEngine.from_engine_args(vllm_engine_config)
+
+        self.vllm_engine = AsyncLLMEngine.from_engine_args(vllm_engine_config)
         self.initialized = True
 
-    def preprocess(self, requests):
-        for req_id, req_data in zip(self.context.request_ids.values(), requests):
-            if req_id not in self.context.cache:
-                data = req_data.get("data") or req_data.get("body")
-                if isinstance(data, (bytes, bytearray)):
-                    data = data.decode("utf-8")
-
-                prompt = data.get("prompt")
-                sampling_params = self._get_sampling_params(req_data)
-                lora_request = self._get_lora_request(req_data)
-                self.context.cache[req_id] = {
-                    "text_len": 0,
-                    "stopping_criteria": self._create_stopping_criteria(req_id),
-                }
-                self.vllm_engine.add_request(
-                    req_id, prompt, sampling_params, lora_request=lora_request
-                )
+    async def handle(self, data, context):
+        start_time = time.time()
 
-        return requests
+        metrics = context.metrics
 
-    def inference(self, input_batch):
-        inference_outputs = self.vllm_engine.step()
-        results = {}
+        data_preprocess = await self.preprocess(data)
+        output = await self.inference(data_preprocess, context)
+        output = await self.postprocess(output)
 
-        for output in inference_outputs:
-            req_id = output.request_id
-            results[req_id] = {
-                "text": output.outputs[0].text[
-                    self.context.cache[req_id]["text_len"] :
-                ],
+        stop_time = time.time()
+        metrics.add_time(
+            "HandlerTime", round((stop_time - start_time) * 1000, 2), None, "ms"
+        )
+        return output
+
+    async def preprocess(self, requests):
+        input_batch = []
+        assert len(requests) == 1, "Expecting batch_size = 1"
```

Comment:
Is this to ensure that the async request handling in the frontend immediately forwards the request to the backend without waiting to form a batch?

Reply:
Correct, in engines like the vllm there is no classical setting as the batch size but number of concurrent requests is influenced by the available memory as well as request length. Therefore, send the request to the engine asap is the best option in my opinion.